File: Bits/sum_of_root_to_leaf_binary_numbers.py
from collections import defaultdict
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        res = []
        stk = [(root, [root.val])]
        while stk:
            cur, path = stk.pop()
            if not cur.left and not cur.right:
                res.append(path)
            if cur.right:
                rpath = path + [cur.right.val]
                stk.append((cur.right, rpath))
            if cur.left:
                lpath = path + [cur.left.val]
                stk.append((cur.left, lpath))
        conv = lambda l: sum(x * 2**i for i, x in enumerate(reversed(l)))
        return sum(map(conv, res))

File: Bits/CPP/sum_of_root_to_leaf_binary_numbers.cpp
#include <cmath>
#include <vector>
#include <utility>
using namespace std;

// Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        int res = 0;
        vector<pair<TreeNode*, vector<int>>> stk = {{root, {root->val}}};
        while (!stk.empty()) {
            TreeNode* node = stk.back().first;
            vector<int> path = stk.back().second;
            stk.pop_back();
            if (!node->left && !node->right) {
                res += conv(path);
            }
            if (node->right) {
                vector<int> rpath = path;
                rpath.push_back(node->right->val);
                stk.push_back({node->right, rpath});
            }
            if (node->left) {
                vector<int> lpath = path;
                lpath.push_back(node->left->val);
                stk.push_back({node->left, lpath});
            }
        }
        return res;
    }

    int conv(vector<int>& bitlist) {
        int res = 0;
        for (int bit : bitlist) res = (res << 1) | bit;
        return res;
    }
};
